src/DotVVM.Framework/Resources/Scripts/utils/evaluator.ts

```diff
@@ -1,18 +1,55 @@
 import { isObservableArray } from "./knockout";
 import { logError } from "./logging";
 
-export function evaluateOnViewModel(context: any, expression: string): any {
-    // todo: reimplement
-    let result;
-    if (context && context.$data) {
-        result = new Function("$context", "with($context) { with ($data) { return " + expression + "; } }")(context);
-    } else {
-        result = new Function("$context", "var $data=$context; with($context) { return " + expression + "; }")(context);
+var shouldBeConvertedFromObservable: (currentLevel: any, remainingParts: string[]) => boolean = (currentLevel: any, remainingParts: string[]) => {
+    if (currentLevel["$data"] == undefined) {
+        return false;
+    }
+
+    for (let i = 0; i < remainingParts.length; i++) {
+        if (remainingParts[i].startsWith("$")) return false;
+    }
+
+    return true;
+};
+
+
+export function evaluateExpression(context: any, expression: string): any {
```

Comment:
the issue is it is exported as "evaluator" API in dotvvm exports. I just removed it. Using evaluate expression is a bad practice anyway. Especcialy now when the function does not do what you would think it would. It only parses specificly validation paths. It does not evaluate any JS expression maybe with the exception of function calls and array indexers.

If anything this should be exported as a part of validation API. But we have agreed to deal with JS validation API in separate PR.

Reply:
Renamed to traverseContext + added few comments that describe the format of the paths.